src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs

```diff
@@ -1787,6 +1787,19 @@ public void WriteMetadataAndIL(PdbWriter nativePdbWriterOpt, Stream metadataStre
 
             if (_debugMetadataOpt != null)
             {
+                // Ensure document table lists files in command line order
+                var documentsBuilder = Module.DebugDocumentsBuilder;
+                var documents = documentsBuilder.DebugDocuments;
+                foreach (var tree in Module.CommonCompilation.SyntaxTrees)
```

Comment:
At this point in the code does this include `SyntaxTree` that come from source generators?

Reply:
Great question. I think the answer is yes because this is emit-phase work which we are only expected to do for the "final" compilation which includes source generator outputs.

It feels like it would be appropriate to have a test here.

Reply:
Definitely need a test - it should also cover that all source generated files are embedded in the PDB.

Reply:
I'm not yet certain how to formulate this test. See this simple source generators test:

https://github.com/dotnet/roslyn/blob/a163778013dba0ad5efb32df5374584b0795b53a/src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs#L113-L140

If I try to imitate the structure of the source generator tests, I will create an original compilation, feed it into the generator driver, then get a final compilation out and try to emit with it. But it seems obvious that the syntax trees for that compilation will include the generated sources, so the test won't really be demonstrating anything useful.

It feels like if we had a higher level test that straight up ran through [`CommonCompiler.Run`](https://github.com/dotnet/roslyn/blob/a163778013dba0ad5efb32df5374584b0795b53a/src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs#L677) we might be able to demonstrate something meaningful, but I'm not sure it's good to spend the time to come up with the automated test just yet.

I imagine that in the near future we might write a test harness that allows us to just take any `Compilation` and attempt to "round-trip" it, by emitting its dll+pdb, gathering the information to recreate it, and then doing so on the spot. Test failure messages could be given by comparing output of `ildasm` or `mdv` or something like that on the before and after binaries.

Reply:
The test could verify that the PDB contains the sources of the generated files. That is, the compilation that comes out of the driver is setup correctly.